sycl/include/sycl/vector.hpp

```diff
@@ -111,24 +97,852 @@ template <typename T> class GetOp {
+    lhs = lhs << shift;
+    return lhs;
+  }
+  friend OpAssignSelfOperandTy &operator>>=(OpAssignSelfOperandTy &lhs,
+                                            int shift) {
+    lhs = lhs >> shift;
+    return lhs;
+  }
+};
+#endif
+
+// We use std::plus<void> and similar to "map" template parameter to an
+// overloaded operator. These three below are missing from `<functional>`.
+struct ShiftLeft {
+  template <class T, class U>
+  constexpr auto operator()(T &&lhs, U &&rhs) const
+      -> decltype(std::forward<T>(lhs) << std::forward<U>(rhs)) {
+    return std::forward<T>(lhs) << std::forward<U>(rhs);
+  }
+};
+struct ShiftRight {
+  template <class T, class U>
+  constexpr auto operator()(T &&lhs,
+                            U &&rhs) const -> decltype(std::forward<T>(lhs) >>
+                                                       std::forward<U>(rhs)) {
+    return std::forward<T>(lhs) >> std::forward<U>(rhs);
+  }
+};
+
+struct UnaryPlus {
+  template <class T>
+  constexpr auto operator()(T &&arg) const -> decltype(+std::forward<T>(arg)) {
+    return +std::forward<T>(arg);
+  }
+};
+
+template <class T>
+static constexpr bool not_fp =
+    !std::is_same_v<T, float> && !std::is_same_v<T, double> &&
+    !std::is_same_v<T, half> && !std::is_same_v<T, ext::oneapi::bfloat16>;
+
+// To provide information about operators availability depending on vec/swizzle
+// element type.
+template <typename Op, typename T>
+inline constexpr bool is_op_available = false;
+
+#define __SYCL_OP_AVAILABILITY(OP, COND)                                       \
+  template <typename T> inline constexpr bool is_op_available<OP, T> = COND;
+
+// clang-format off
+__SYCL_OP_AVAILABILITY(std::plus<void>          , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::minus<void>         , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::multiplies<void>    , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::divides<void>       , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::modulus<void>       , !detail::is_byte_v<T> && not_fp<T>)
+
+__SYCL_OP_AVAILABILITY(std::bit_and<void>       , not_fp<T>)
+__SYCL_OP_AVAILABILITY(std::bit_or<void>        , not_fp<T>)
+__SYCL_OP_AVAILABILITY(std::bit_xor<void>       , not_fp<T>)
+
+__SYCL_OP_AVAILABILITY(std::equal_to<void>      , true)
+__SYCL_OP_AVAILABILITY(std::not_equal_to<void>  , true)
+__SYCL_OP_AVAILABILITY(std::less<void>          , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::greater<void>       , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::less_equal<void>    , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::greater_equal<void> , !detail::is_byte_v<T>)
+
+__SYCL_OP_AVAILABILITY(std::logical_and<void>   , !detail::is_byte_v<T>)
+__SYCL_OP_AVAILABILITY(std::logical_or<void>    , !detail::is_byte_v<T>)
+
+__SYCL_OP_AVAILABILITY(ShiftLeft     , !detail::is_byte_v<T> && not_fp<T>)
+__SYCL_OP_AVAILABILITY(ShiftRight    , !detail::is_byte_v<T> && not_fp<T>)
```

Comment:
Nit; If we're turning off the formatter for the sake of readability here, why not align these with the others?

Reply:
I think it's the same either way... Will update in the next upload.